### PR TITLE
Make the `serviceRequest` function more flexible.

### PR DESCRIPTION
**Pull request by nh758**
`serviceRequest()` now supports options to give more control to our retries and timeout. Also made the callback optional, allowing us to use it as a promise. I include unit tests, testing it works in the different way it can be called, including the existing format.

### abRequestService.serviceRequest(key, data, [options], [cb]) ⇒ <code>Promise</code>
Send a request to another micro-service using the cote protocol. Accepts an
optional callback, but also returns a promise.

**Kind**: instance method of [<code>ABRequestService</code>](#ABRequestService)  
**Returns**: <code>Promise</code> - resolves with the response from the service  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| key | <code>string</code> |  | the service handler's key we are sending a request to. |
| data | <code>json</code> |  | the data packet to send to the service. |
| [options] | <code>object</code> |  | options |
| [options.timeout] | <code>number</code> | <code>5000</code> | ms to wait before timing out |
| [options.maxAttempts] | <code>number</code> | <code>5</code> | how many times to try the request if  it fails |
| [options.longRequest] | <code>boolean</code> | <code>false</code> | timeout after 90 seconds, will be ignored if timeout was set |
| [cb] | <code>function</code> |  | node.js style callback(err, result) for when the response is received. |

**Example**  
```js
// async/await
try {
   let result = await serviceRequest(key, data);
} catch (err) {}
// promise
serviceRequest(key, data, opts).then((result) => {}).catch((err) => {})
// callback
serviceRequest(key, data, opts, (result, err) => {})
// or
serviceRequest(key, data, (result, err) => {})
```

### CODE REVIEW

File: utils/serviceRequest.js [hunk 1]
```diff
@@ -20,62 +20,98 @@ const domainRequesters = {
...
+                  "Warning: serviceRequest() now supports an options parameter `serviceRequest(key, data, options = {}, callback?)`. Please refactor longRequest to options",
+            }
+         );
+         options.longRequest = data.longRequest;
+         delete data.longRequest;
       }
-      let countRequest = 0;
-      const longRequest = data.longRequest ?? false;
-      delete data.longRequest; // The service does not need this passed.
+      // console.log("options", options);
```
Comment (jinzsom) [inline utils/serviceRequest.js]: LGTM but I don't think we need this?

**Review by wongpratan — APPROVED** [review]

**Review by jinzsom — APPROVED** [review]

File: utils/serviceRequest.js [hunk 2]
```diff
@@ -20,62 +20,98 @@ const domainRequesters = {
...
+    * @returns {Promise} resolves with the response from the service
+    * @example
+    * // async/await
+    * try {
+    *    let result = await request(key, data);
+    * } catch (err) {}
+    * // promise
+    * request(key, data, opts).then((result) => {}).catch((err) => {})
+    * // callback
+    * request(key, data, opts, (result, err) => {})
```
Comment (Hiro-Nakamura) [inline utils/serviceRequest.js]: minor issue: the callback format is  `(err, result)`

File: utils/reqService.js [hunk 3]
```diff
@@ -737,18 +737,34 @@ class ABRequestService {
...
+    * @returns {Promise} resolves with the response from the service
+    * @example
+    * // async/await
+    * try {
+    *    let result = await serviceRequest(key, data);
+    * } catch (err) {}
+    * // promise
+    * serviceRequest(key, data, opts).then((result) => {}).catch((err) => {})
+    * // callback
+    * serviceRequest(key, data, opts, (result, err) => {})
```
Comment (Hiro-Nakamura) [inline utils/reqService.js]: `(err, result)`

**Review by Hiro-Nakamura — APPROVED** [review]
This looks good. Nice job.

one minor issue in your documentation: the callback format should be `(err, results)`

This has to be one of the most well documented PRs I've checked over.
And additional points for including the unit tests +1 

But have to say you are setting the bar pretty high ...

**Pull request merged**